Source/options.cpp

```diff
@@ -1419,6 +1418,249 @@ uint32_t KeymapperOptions::KeyForAction(string_view actionName) const
+	auto notFoundIt = sgOptions.Padmapper.buttonToButtonName.end();
+	if ((value.modifier != ControllerButton_NONE && modifierNameIt == notFoundIt) || (value.button != ControllerButton_NONE && buttonNameIt == notFoundIt)) {
+		// Ignore invalid button combos
+		return false;
+	}
+
+	// Remove old button combo
+	if (boundInput.button != ControllerButton_NONE) {
+		boundInput = {};
+		boundInputDescription = "";
+	}
+
+	// Add new button combo
+	if (value.button != ControllerButton_NONE) {
+		boundInput = value;
+		boundInputDescription = buttonNameIt->second;
+
+		if (modifierNameIt != notFoundIt) {
+			boundInputDescription = StrCat(modifierNameIt->second, "+", boundInputDescription);
+		}
+	}
+
+	return true;
+}
+
+void PadmapperOptions::AddAction(string_view key, const char *name, const char *description, ControllerButtonCombo defaultInput, std::function<void()> actionPressed, std::function<void()> actionReleased, std::function<bool()> enable, unsigned index)
+{
+	actions.push_back(std::unique_ptr<Action>(new Action(key, name, description, defaultInput, std::move(actionPressed), std::move(actionReleased), std::move(enable), index)));
+}
+
+void PadmapperOptions::ButtonPressed(ControllerButton button)
+{
+	// To give preference to button combinations,
+	// first pass ignores mappings where no modifier is bound
+	for (auto it = actions.begin(); it != actions.end(); ++it) {
+		Action &action = **it;
+		ControllerButtonCombo combo = action.boundInput;
+		if (combo.modifier == ControllerButton_NONE)
+			continue;
+		if (button != combo.button)
+			continue;
+		if (!IsControllerButtonPressed(combo.modifier))
+			continue;
+		if (action.enable && !action.enable())
+			continue;
+		if (action.actionPressed)
+			action.actionPressed();
+		SuppressedButton = combo.modifier;
+		buttonToReleaseAction.insert_or_assign(combo.button, action);
+		return;
+	}
+
+	for (auto it = actions.begin(); it != actions.end(); ++it) {
+		Action &action = **it;
+		ControllerButtonCombo combo = action.boundInput;
+		if (combo.modifier != ControllerButton_NONE)
+			continue;
+		if (button != combo.button)
+			continue;
+		if (action.enable && !action.enable())
+			continue;
+		if (action.actionPressed)
+			action.actionPressed();
+		SuppressedButton = combo.modifier;
+		buttonToReleaseAction.insert_or_assign(combo.button, action);
+		return;
+	}
+}
+
+void PadmapperOptions::ButtonReleased(ControllerButton button)
+{
+	auto it = buttonToReleaseAction.find(button);
+	if (it == buttonToReleaseAction.end())
+		return; // Ignore unmapped buttons.
+
+	const Action &action = it->second.get();
+
+	// Check that the action can be triggered.
+	if (!action.actionReleased || (action.enable && !action.enable()))
+		return;
+
+	action.actionReleased();
+	buttonToReleaseAction.erase(button);
+}
+
+string_view PadmapperOptions::InputNameForAction(string_view actionName) const
+{
+	for (const auto &action : actions) {

```

Comment:
A reference to a pointer seems like a typo

```suggestion
	for (const Action *action : actions) {
```